### PR TITLE
Add I2CMultiplexer in generel and the TCA9548A in special

### LINKED ISSUE #807: Support for i2c multiplexer tca9548a
tca9548a is an i2c multiplexer, allowing to create 8 i2c busses from a single bus

Really useful when you can't change the remote i2c address and you have to connect multiple devices with the same address

Implementation seems pretty simple, to talk to one bus,you just have to send a byte with the bus number and following communicatons are routed automatically

### PR DESCRIPTION
**Pull request by andreashergert1984**
## Description:
This features adds the possibility to use I2C multiplexer. 
It adds a I2CMultiplexer Class (derived from I2CDevice) to add more known multiplexer.
The famous TCA9548A is included.
on I2CDevices Parameters to activate the channel on a multiplexer was added.

**Related feature request :** Closes https://github.com/esphome/feature-requests/issues/807

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#897

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
Just a thought, maybe the multiplexer code could be wrapped in an `#ifdef I2C_MULTIPLEXER` so that it does not get compiled in when people are not using it?

**Comment by andreashergert1984** [comment]
> Just a thought, maybe the multiplexer code could be wrapped in an `#ifdef I2C_MULTIPLEXER` so that it does not get compiled in when people are not using it?

Good idea, but i wouldn´t know how. At least in void set_i2c_multiplexer(I2CMultiplexer *multiplexer, uint8_t channel); it will be dependend.

**Review by jesserockz — COMMENTED** [review]
Please add yourself to CODEOWNERS in `esphome/components/tca9548a/__init__.py` and run `script/build_codeowners.py`

File: esphome/components/i2c/i2c.h [hunk 1]
```diff
@@ -168,6 +169,11 @@ class I2CDevice {
   /// Manually set the i2c address of this device.
   void set_i2c_address(uint8_t address);
 
+  /// Manually set the i2c multiplexer of this device.
+  void set_i2c_multiplexer(I2CMultiplexer *multiplexer, uint8_t channel);
+
+  // Checks for multiplexer set and set channel
+  void check_multiplexer();
```
Comment (badbadc0ffee) [inline esphome/components/i2c/i2c.h:176]: Does this need to / make sense to be public?

File: esphome/components/tca9548a/tca9548a.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+class TCA9548AComponent : public Component, public i2c::I2CMultiplexer {
+ public:
+  void set_scan(bool scan) { scan_ = scan; }
+  void setup() override;
+  void dump_config() override;
+  void update();
+  bool set_channel(uint8_t channelno) override;
+
+ protected:
+  bool scan_;
```
Comment (badbadc0ffee) [inline esphome/components/tca9548a/tca9548a.h:18]: This appears to be unused?

It looks like scanning the I2C bus will not work with this MUX in place.
The component and / or the base I2C component should issue a warnig when a configuration demands scanning of the I2C bus while a mux is also configured.

File: esphome/components/i2c/i2c.h (hunk 1 above)
Comment (badbadc0ffee) [reply esphome/components/i2c/i2c.h:176]: Could this be made `const`?

Comment (andreashergert1984) [reply esphome/components/i2c/i2c.h:176]: > Does this need to / make sense to be public?
set_i2c_multiplexer: this is called in the  register_i2c_device.. so from "outside".. 
check_multiplexer is not protected.

about const i´m not sure as it potentially modifies multiple multiplexers (as tested locally)

File: esphome/components/tca9548a/tca9548a.h (hunk 2 above)
Comment (andreashergert1984) [reply esphome/components/tca9548a/tca9548a.h:18]: scanning (incl. all busses attached) can be activated in the configuration.
so it is easy to see on which channel which addresses are visible.. and for sure.. it switches channel and uses the scan-function of the bus.. so all devices currently "connected" are seen.

Comment (andreashergert1984) [reply esphome/components/tca9548a/tca9548a.h:18]: fixed: added the check for the scan_ ...

**Comment by andreashergert1984** [comment]
added to test1.yaml

File: esphome/components/tca9548a/__init__.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import esphome.config_validation as cv
+from esphome.components import i2c
+from esphome.const import CONF_ID, CONF_SCAN
+
+CODEOWNERS = ['@andreashergert1984']
+
+DEPENDENCIES = ['i2c']
+
+tca9548a_ns = cg.esphome_ns.namespace('tca9548a')
+TCA9548AComponent = tca9548a_ns.class_('TCA9548AComponent', cg.PollingComponent, i2c.I2CDevice)
```
Comment (jesserockz) [inline esphome/components/tca9548a/__init__.py:11]: Shouldn't this be `I2CMultiplexer`?

```suggestion
TCA9548AComponent = tca9548a_ns.class_('TCA9548AComponent', cg.PollingComponent, i2c.I2CMultiplexer)
```

File: esphome/components/i2c/__init__.py [hunk 4]
```diff
@@ -20,6 +22,11 @@
     cv.Optional(CONF_SCAN, default=True): cv.boolean,
 }).extend(cv.COMPONENT_SCHEMA)
 
+I2CMULTIPLEXER_SCHEMA = cv.Schema({
+    cv.Required(CONF_ID): cv.use_id(I2CDevice),
```
Comment (jesserockz) [inline esphome/components/i2c/__init__.py:26]: We want this to specifically be a multiplexer id.

```suggestion
    cv.Required(CONF_ID): cv.use_id(I2CMultiplexer),
```

File: esphome/components/i2c/i2c.cpp [hunk 5]
```diff
@@ -178,28 +178,47 @@ bool I2CComponent::write_byte_16(uint8_t address, uint8_t a_register, uint16_t d
 }
 
 void I2CDevice::set_i2c_address(uint8_t address) { this->address_ = address; }
+void I2CDevice::set_i2c_multiplexer(I2CMultiplexer *multiplexer, uint8_t channel) {
+  ESP_LOGVV(TAG, "    Setting Multiplexer %p for channel %d", multiplexer, channel);
+  this->multiplexer_ = multiplexer;
+  this->channel_ = channel;
+}
+void I2CDevice::check_multiplexer_() {
+  if (this->multiplexer_) {
```
Comment (jesserockz) [inline esphome/components/i2c/i2c.cpp:187]: To keep consistency with other code that checks for `nullptr`

```suggestion
  if (this->multiplexer_ != nullptr) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
A couple of minor changes. 

I just want to make sure we are not going to break I2C for everything once this goes stable.

File: esphome/components/i2c/__init__.py [hunk 6]
```diff
@@ -2,13 +2,15 @@
...
 from esphome.const import CONF_FREQUENCY, CONF_ID, CONF_SCAN, CONF_SCL, CONF_SDA, CONF_ADDRESS, \
-    CONF_I2C_ID
+    CONF_I2C_ID, CONF_CHANNEL, CONF_MULTIPLEXER
 from esphome.core import coroutine, coroutine_with_priority
 
 CODEOWNERS = ['@esphome/core']
 i2c_ns = cg.esphome_ns.namespace('i2c')
 I2CComponent = i2c_ns.class_('I2CComponent', cg.Component)
 I2CDevice = i2c_ns.class_('I2CDevice')
+I2CMultiplexer = i2c_ns.class_('I2CMultiplexer')
```
Comment (jesserockz) [inline esphome/components/i2c/__init__.py:12]: I don't understand the amazing codegen enough to argue as to why the parent classes are always listed here as per the c++ code...

```suggestion
I2CMultiplexer = i2c_ns.class_('I2CMultiplexer', I2CDevice)
```

**Review by jesserockz — APPROVED** [review]
I think this looks good now, but I am going to wait for another approval before merging it in.

File: esphome/components/tca9548a/tca9548a.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      this->set_channel(i);
+      this->parent_->dump_config();
+    }
+  }
+}
+
+bool TCA9548AComponent::set_channel(uint8_t channelno) {
+  uint8_t channelbyte = 1 << channelno;
+  this->write_byte(0x70, channelbyte);
+  return true;
```
Comment (glmnet) [inline esphome/components/tca9548a/tca9548a.cpp:33]: why making this always return true

File: tests/test1.yaml [hunk 8]
```diff
@@ -1830,6 +1830,11 @@ cover:
 
 debug:
 
+tca9548a:
+  - address: 0x70
+    id: multiplex0
+    scan: True
+
```
Comment (glmnet) [inline tests/test1.yaml:1837]: may be also adding a device to use the multiplexer could catch unexpected issues

File: esphome/components/tca9548a/tca9548a.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+      ESP_LOGCONFIG(TAG, "Activating channel: %d", i);
+      this->set_channel(i);
+      this->parent_->dump_config();
+    }
+  }
+}
+
+bool TCA9548AComponent::set_channel(uint8_t channelno) {
+  uint8_t channelbyte = 1 << channelno;
+  this->write_byte(0x70, channelbyte);
```
Comment (glmnet) [inline esphome/components/tca9548a/tca9548a.cpp:32]: wouldn't it make sense to keep track of the active channel and change only if necessary?

File: esphome/components/i2c/__init__.py [hunk 10]
```diff
@@ -43,6 +50,7 @@ def i2c_device_schema(default_address):
     """
     schema = {
         cv.GenerateID(CONF_I2C_ID): cv.use_id(I2CComponent),
+        cv.Optional(CONF_MULTIPLEXER): cv.Schema(I2CMULTIPLEXER_SCHEMA),
```
Comment (glmnet) [inline esphome/components/i2c/__init__.py:53]: I2CMULTIPLEXER_SCHEMA is already a schema
```suggestion
        cv.Optional(CONF_MULTIPLEXER): I2CMULTIPLEXER_SCHEMA,
```

File: esphome/components/i2c/i2c.cpp [hunk 11]
```diff
@@ -178,28 +178,47 @@ bool I2CComponent::write_byte_16(uint8_t address, uint8_t a_register, uint16_t d
...
 
 void I2CDevice::set_i2c_address(uint8_t address) { this->address_ = address; }
+void I2CDevice::set_i2c_multiplexer(I2CMultiplexer *multiplexer, uint8_t channel) {
+  ESP_LOGVV(TAG, "    Setting Multiplexer %p for channel %d", multiplexer, channel);
+  this->multiplexer_ = multiplexer;
+  this->channel_ = channel;
+}
+void I2CDevice::check_multiplexer_() {
+  if (this->multiplexer_ != nullptr) {
+    ESP_LOGVV(TAG, "Multiplexer present setting channel to channel %d", this->channel_);
```
Comment (glmnet) [inline esphome/components/i2c/i2c.cpp:188]:
```suggestion
    ESP_LOGVV(TAG, "Multiplexer setting channel to %d", this->channel_);
```

**Review by glmnet — CHANGES_REQUESTED** [review]
I agree with @jesserockz this should be less intrusive, when multiplexers are not used which is the norm.

There are a few options, 
* c++ inheritance, this is what I would have done first, easy to understand but adds inheritance overhead (vtable etc)
* static polymorphism might be used but adds much complexity, also if using a device both with a multiplexer and without then it might not be good idea
* this one is nice unless you don't need it, specifically adds 
  - a check on every call to read / write
  - a multiplexer and channel field on each i2c device
  
  the above doesn't seem much but there are already people hitting the memory limits of the esps

please look for `cg.add_define(` in the code and add `USE_I2C_MULTIPLEXER`

**Comment by dougle03** [comment]
Any plans for this to be released anytime soon?

**Comment by andreashergert1984** [comment]
> Any plans for this to be released anytime soon?

as here is as requested some work to do .. to not effect non-multiplexer users as far as possible.. and a lack of time on my side.. it will need some further. time.. feel free to make those changes and make a PR. ;)

**Comment by dougle03** [comment]
> > Any plans for this to be released anytime soon?
> 
> as here is as requested some work to do .. to not effect non-multiplexer users as far as possible.. and a lack of time on my side.. it will need some further. time.. feel free to make those changes and make a PR. ;)

I really wish I had the skillset to do this, alas I wait in eternal hope that you find the time in the near future. best

**Comment by Casino** [comment]
> 
> 
> > > Any plans for this to be released anytime soon?
> > 
> > 
> > as here is as requested some work to do .. to not effect non-multiplexer users as far as possible.. and a lack of time on my side.. it will need some further. time.. feel free to make those changes and make a PR. ;)
> 
> I really wish I had the skillset to do this, alas I wait in eternal hope that you find the time in the near future. best

Yes, I'm also looking forward to see this feature merged! Unfortunatelly, best I can do is to test the multiplexer locally. Until now it works quite fine with my esp8266. Thanks a lot for your work! Best whishes

File: esphome/components/i2c/i2c.h [hunk 12]
```diff
@@ -280,9 +282,18 @@ class I2CDevice {
   bool write_byte_16(uint8_t a_register, uint16_t data);
 
  protected:
+  // Checks for multiplexer set and set channel
+  void check_multiplexer_();
   uint8_t address_{0x00};
   I2CComponent *parent_{nullptr};
+  I2CMultiplexer *multiplexer_{nullptr};
+  uint8_t channel_;
```
Comment (glmnet) [inline esphome/components/i2c/i2c.h:290]: Wrap in #ifdef to save memory

**Review by glmnet — CHANGES_REQUESTED** [review]

File: esphome/components/tca9548a/tca9548a.cpp (hunk 7 above)
Comment (andreashergert1984) [reply esphome/components/tca9548a/tca9548a.cpp]: removed ;)

File: tests/test1.yaml (hunk 8 above)
Comment (andreashergert1984) [reply tests/test1.yaml:2000]: check ... added a second one "after" the first.. like in my real-life application

File: esphome/components/tca9548a/tca9548a.cpp (hunk 9 above)
Comment (andreashergert1984) [reply esphome/components/tca9548a/tca9548a.cpp]: done

File: esphome/components/i2c/i2c.h (hunk 12 above)
Comment (glmnet) [reply esphome/components/i2c/i2c.h:290]: You will need to `#include "esphome/core/defines.h"`

Comment (andreashergert1984) [reply esphome/components/i2c/i2c.h:293]: done

**Review by glmnet — APPROVED** [review]

**Pull request merged**